### PR TITLE
Add broadcast operations via WebSocket

### PR DESCRIPTION
**Pull request by aoudiamoncef**
* [x] document all added functions
* [ ] try in sandbox /simulation/labnet
* [x] unit tests on the added/changed features
  * [x] make tests compile
  * [x] make tests pass 
* [x] add logs allowing easy debugging in case the changes caused problems
* [x] if the API has changed, update the API specification

### CODE REVIEW

File: massa-node/base_config/config.toml [hunk 1]
```diff
@@ -132,6 +132,8 @@
     max_operations_propagation_time = 32000
     # time threshold after which endorsement are not propagated
     max_endorsements_propagation_time = 48000
+    # operations sender(channel) capacity
+    ws_operations_capacity = 128
```
Comment (AurelienFT) [inline massa-node/base_config/config.toml:136]: That's a really low number as we have multiple thousands of op in 1 second and you can receive batch that are way more huge then 128. I think you can put 5_000 at least

File: massa-ws/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+use jsonrpsee::{core::error::SubscriptionClosed, SubscriptionSink};
+use serde::Serialize;
+use tokio::sync::broadcast::Sender;
+use tokio_stream::wrappers::BroadcastStream;
+
+/// Re-exports libraries to not require any additional
+/// dependencies to be explicitly added on the module(s) side.
+#[doc(hidden)]
+pub mod __reexports {
```
Comment (AurelienFT) [inline massa-ws/src/lib.rs:15]: Never saw this way of re-exporting. The idiomatic way (I believe) is to `pub use` the part you want to re-export. I have made this in my previous PR : https://github.com/massalabs/massa/blob/testnet_18/massa-execution-exports/src/lib.rs#L56 (I re-export some massa-sc-runtime stuff)

File: massa-protocol-worker/src/protocol_worker.rs [hunk 3]
```diff
@@ -1080,6 +1100,11 @@ impl ProtocolWorker {
 
         Ok(())
     }
+
+    /// New produced operations.
+    pub(crate) fn subscribe_new_operations(&self, sink: SubscriptionSink) {
+        broadcast_via_ws(self.ws_config.operation_sender.clone(), sink);
```
Comment (AurelienFT) [inline massa-protocol-worker/src/protocol_worker.rs:1106]: I don't think making a function for a line of code it's worth it.

File: massa-consensus-worker/src/controller.rs [hunk 4]
```diff
@@ -301,29 +298,14 @@ impl ConsensusController for ConsensusControllerImpl {
     }
 
     fn subscribe_new_blocks_headers(&self, sink: SubscriptionSink) {
```
Comment (AurelienFT) [inline massa-consensus-worker/src/controller.rs:300]: While you are refactoring I don't really understand what this functions are doing. Is not suppose to be the API that start the thread when it receive someone subscribing ? And the modules only broadcast ? 

So in your module you just need to know about a tokio sender and the ws related stuff is contained in API.

File: massa-node/base_config/openrpc.json [hunk 5]
```diff
@@ -1125,7 +1125,33 @@
...
+                }
+            ],
+            "params": [],
+            "result": {
+                "schema": {
+                    "$ref": "#/components/schemas/Operation"
+                },
+                "name": "Operation"
+            },
+            "name": "subscribe_new_filled_blocks",
```
Comment (AurelienFT) [inline massa-node/base_config/openrpc.json:1152]: I think the name isn't good here.

**Review by AurelienFT — COMMENTED** [review]
Good job just a few concern about code splitting.

File: massa-node/base_config/config.toml (hunk 1 above)
Comment (aoudiamoncef) [reply massa-node/base_config/config.toml:136]: Yes, I knew that you'll give us a good one

File: massa-ws/src/lib.rs (hunk 2 above)
Comment (aoudiamoncef) [reply massa-ws/src/lib.rs:15]: IMHO, this way is used to share an entire dependency. But in our case may be a `pub use` could be used.

Comment (AurelienFT) [reply massa-ws/src/lib.rs:15]: I don't think is necessary to share a whole module yes.

File: massa-protocol-worker/src/protocol_worker.rs (hunk 3 above)
Comment (aoudiamoncef) [reply massa-protocol-worker/src/protocol_worker.rs:1106]: this call will be replaced by: 

```rust
/// Brodcast the stream(sender) content via a WebSocket
pub fn broadcast_via_ws<T: Serialize + Send + Clone + 'static>(
    sender: Sender<T>,
    mut sink: SubscriptionSink,
) {
    let rx = BroadcastStream::new(sender.subscribe());
    tokio::spawn(async move {
        match sink.pipe_from_try_stream(rx).await {
            SubscriptionClosed::Success => {
                sink.close(SubscriptionClosed::Success);
            }
            SubscriptionClosed::RemotePeerAborted => (),
            SubscriptionClosed::Failed(err) => {
                sink.close(err);
            }
        };
    });
}
```

Comment (aoudiamoncef) [reply massa-protocol-worker/src/protocol_worker.rs:1106]: If we do this, the `massa-ws` module will be removed.

File: massa-consensus-worker/src/controller.rs (hunk 4 above)
Comment (aoudiamoncef) [reply massa-consensus-worker/src/controller.rs:300]: May be the module `massa-ws` could be removed. see the comment above please

File: massa-node/base_config/openrpc.json (hunk 5 above)
Comment (aoudiamoncef) [reply massa-node/base_config/openrpc.json:1152]: done

File: massa-protocol-worker/src/protocol_worker.rs (hunk 3 above)
Comment (AurelienFT) [reply massa-protocol-worker/src/protocol_worker.rs:1106]: I'm saying that `broadcast_via_ws` is already a function and making the function `subscribe_new_operations` that only have one line of code seems to be an overhead.

File: massa-consensus-worker/src/controller.rs (hunk 4 above)
Comment (AurelienFT) [reply massa-consensus-worker/src/controller.rs:300]: It's not related to the comment above. It's about using the websocket things only in API and only have sender in the other modules.

File: massa-node/base_config/config.toml (hunk 1 above)
Comment (aoudiamoncef) [reply massa-node/base_config/config.toml:136]: done

File: massa-ws/src/lib.rs (hunk 2 above)
Comment (aoudiamoncef) [reply massa-ws/src/lib.rs]: done

File: massa-protocol-worker/src/protocol_worker.rs (hunk 3 above)
Comment (aoudiamoncef) [reply massa-protocol-worker/src/protocol_worker.rs]: done

File: massa-consensus-worker/src/controller.rs (hunk 4 above)
Comment (aoudiamoncef) [reply massa-consensus-worker/src/controller.rs]: done

File: massa-node/base_config/openrpc.json [hunk 6]
```diff
@@ -1125,7 +1125,33 @@
...
+            ],
+            "params": [],
+            "result": {
+                "schema": {
+                    "$ref": "#/components/schemas/Operation"
+                },
+                "name": "Operation"
+            },
+            "name": "subscribe_new_operations",
+            "summary": "New produced operations",
```
Comment (AurelienFT) [inline massa-node/base_config/openrpc.json:1153]:
```suggestion
            "summary": "Subscribe to new received operations",
```

File: massa-node/base_config/openrpc.json [hunk 7]
```diff
@@ -1234,6 +1260,42 @@
...
+            ],
+            "result": {
+                "schema": {
+                    "type": "boolean"
+                },
+                "name": "unsubscribe result",
+                "description": "unsubscribe success message"
+            },
+            "name": "unsubscribe_new_operations",
+            "summary": "Unsubscribe from new produced operations",
```
Comment (AurelienFT) [inline massa-node/base_config/openrpc.json:1297]:
```suggestion
            "summary": "Unsubscribe from new received operations",
```

File: massa-node/base_config/openrpc.json [hunk 8]
```diff
@@ -1234,6 +1260,42 @@
...
+            "description": "Unsubscribe from new produced operations."
```
Comment (AurelienFT) [inline massa-node/base_config/openrpc.json:1298]:
```suggestion
            "description": "Unsubscribe from new received operations."
```

File: massa-node/base_config/openrpc.json [hunk 9]
```diff
@@ -1125,7 +1125,33 @@
...
+            "params": [],
+            "result": {
+                "schema": {
+                    "$ref": "#/components/schemas/Operation"
+                },
+                "name": "Operation"
+            },
+            "name": "subscribe_new_operations",
+            "summary": "New produced operations",
+            "description": "New produced operations."
```
Comment (AurelienFT) [inline massa-node/base_config/openrpc.json:1154]:
```suggestion
            "description": "New received operations."
```

File: massa-api/src/api_trait.rs [hunk 10]
```diff
@@ -34,4 +34,12 @@ pub trait MassaApi {
 		item = FilledBlock
 	)]
     fn subscribe_new_filled_blocks(&self);
+
+    /// New produced operations.
```
Comment (AurelienFT) [inline massa-api/src/api_trait.rs:38]:
```suggestion
    /// New received operations.
```

**Review by AurelienFT — COMMENTED** [review]
Much better separation and organisation.

Do we also want to announce the operations that are send in the API of this node ? Because currently we only take the ones that come from the network but we don't take the one sent by user through the API.

I think this could be in a separate issue because it implies more concerns, can you just specify in the description of openrpc that we don't broadcast op send through the node API on this node. And open an issue to do it later on.

Thank you

**Comment by AurelienFT** [comment]
The tests are not compiling.

**Comment by aoudiamoncef** [comment]
We could add existing

> Much better separation and organisation.
>
> Do we also want to announce the operations that are send in the API of this node ? Because currently we only take the ones that come from the network but we don't take the one sent by user through the API.

 IMHO, we could do It rapidly, we have to push new RAW operations but without any control...

> I think this could be in a separate issue because it implies more concerns, can you just specify in the description of openrpc that we don't broadcast op send through the node API on this node. And open an issue to do it later on.

W'll explicitly do it.

**Comment by AurelienFT** [comment]
> IMHO, we could do It rapidly, we have to push new RAW operations but without any control...

That's the problem on my mind we don't have verifications and the node operator already have them if he catch them in the API. Let's see in the near future to merge this for now.

**Comment by aoudiamoncef** [comment]
@AurelienFT , tests seems to be green. We could continue

**Review by AurelienFT — APPROVED** [review]

**Pull request merged**